src/sources/syslog.rs

```diff
@@ -219,59 +223,108 @@ impl SyslogDecoder {
         // |
         // | ASCII decimal number of unknown length
 
-        if let Some(i) = src.iter().position(|&b| b == b' ') {
-            let len: usize = std::str::from_utf8(&src[..i])
-                .map_err(|_| ())
-                .and_then(|num| num.parse().map_err(|_| ()))
-                .map_err(|_| {
-                    LinesCodecError::Io(io::Error::new(
-                        io::ErrorKind::InvalidData,
-                        "Unable to decode message len as number",
-                    ))
-                })?;
-
-            let from = i + 1;
-            let to = from + len;
-
-            if let Some(msg) = src.get(from..to) {
-                let s = std::str::from_utf8(msg)
-                    .map_err(|_| {
-                        LinesCodecError::Io(io::Error::new(
+        let space_pos = src.iter().position(|&b| b == b' ');
+
+        // If we are discarding, discard to the next newline.
+        let newline_pos = src.iter().position(|&b| b == b'\n');
+
+        match (self.is_discarding, newline_pos, space_pos) {
+            (true, Some(offset), _) => {
+                // When discarding we keep discarding to the next newline.
+                src.advance(offset + 1);
+                self.is_discarding = false;
+                self.in_octet = false;
+                Err(LinesCodecError::Io(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Frame length limit exceeded",
+                )))
+            }
+            (true, None, _) => {
+                // There is no newline in this frame. Advance as far as we can to
+                // discard the entire frame.
+                src.advance(src.len());
+                Ok(None)
+            }
+            (false, _, Some(space_pos)) => {
+                let len: usize = match std::str::from_utf8(&src[..space_pos])
+                    .map_err(|_| ())
+                    .and_then(|num| num.parse().map_err(|_| ()))
+                {
+                    Ok(len) => len,
```

Comment:
This still isn't quite right. We aren't handling the case where you specify a length > maximum length, or a length for a message that is split over multiple frames.

Reply:
We should be now.